src/autoTune.ml

```diff
@@ -0,0 +1,434 @@
+
+
+(*Functions for determining if the congruence analysis should be enabled *)
+let isExtern = function
+  | Extern -> true
+  | _ -> false
+
+let rec setCongruenceRecursive fd depth neigbourFunction =
+  if depth >= 0 then (
+    fd.svar.vattr <- addAttributes (fd.svar.vattr) [Attr ("goblint_precision",[AStr "congruence"])];
+    FunctionSet.iter
+      (fun vinfo ->
+         print_endline ("    " ^ vinfo.vname);
+         setCongruenceRecursive (Cilfacade.find_varinfo_fundec vinfo) (depth -1) neigbourFunction
+      )
+      (FunctionSet.filter (*for extern and builtin functions there is no function definition in CIL*)
+         (fun x -> not (isExtern x.vstorage || String.starts_with ~prefix:"__builtin" x.vname))
+         (neigbourFunction fd.svar)
+      )
+    ;
+  )
+
+exception ModFound
+class modVisitor = object
+  inherit nopCilVisitor
+
+  method! vexpr = function
+    | BinOp (Mod,_,_,_) ->
+      raise ModFound;
+    | _ -> DoChildren
+end
+
+class modFunctionAnnotatorVisitor = object
+  inherit nopCilVisitor
+
+  method! vfunc fd =
+    let thisVisitor = new modVisitor in
+    try ignore (visitCilFunction thisVisitor fd) with
+    | ModFound ->
+      print_endline ("function " ^ (CilType.Fundec.show fd) ^" uses mod, enable congruence domain recursively for:");
+      print_endline ("  \"down\":");
+      setCongruenceRecursive fd 6 calledFunctions;
+      print_endline ("  \"up\":");
+      setCongruenceRecursive fd 3 callingFunctions;
+      ;
+      SkipChildren
+end
+
+let addModAttributes file =
+  set_bool "annotation.int.enabled" true;
+  let thisVisitor = new modFunctionAnnotatorVisitor in
+  ignore (visitCilFileSameGlobals thisVisitor file)
+
+
+let disableIntervalContextsInRecursiveFunctions () =
+  ResettableLazy.force functionCallMaps |> fun (x,_,_) -> x |> FunctionCallMap.iter (fun f set ->
+      (*detect direct recursion and recursion with one indirection*)
+      if FunctionSet.mem f set || (not @@ FunctionSet.disjoint (calledFunctions f) (callingFunctions f)) then (
+        print_endline ("function " ^ (f.vname) ^" is recursive, disable interval context");
+        f.vattr <- addAttributes (f.vattr) [Attr ("goblint_context",[AStr "base.no-interval"])];
+      )
+    )
+
+(*If only one thread is used in the program, we can disable most thread analyses*)
+(*The exceptions are analyses that are depended on by others: base -> mutex -> mutexEvents, access*)
+(*escape is also still enabled, because otherwise we get a warning*)
+(*does not consider dynamic calls!*)
+
+let notNeccessaryThreadAnalyses = ["deadlock"; "maylocks"; "symb_locks"; "thread"; "threadflag"; "threadid"; "threadJoins"; "threadreturn"]
+let reduceThreadAnalyses () =
+  let hasThreadCreate () =
+    ResettableLazy.force functionCallMaps
+    |> fun (_,x,_) -> x  (*every function that is called*)
+                      |> FunctionCallMap.exists
```

Comment:
It s the one used by icp-indent, because the second chain of |> is inside the function. I thought that at some point I  had put line 142 in brackets to change it into the mostly equivalent version where it is a single chain of |>. Should I just override icp-indent or do it like that again?

Reply:
I probably re-introduced it when running `ocp-indent`. Feel free to change it again, `ocp-indent` sometimes does strange things, and we are fine with some local deviations from what it suggest.